optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/incremental/IncrementalScoreDirector.java

```diff
@@ -42,6 +42,12 @@ public IncrementalScoreDirector(IncrementalScoreDirectorFactory scoreDirectorFac
                                     boolean constraintMatchEnabledPreference,
                                     IncrementalScoreCalculator<Solution_> incrementalScoreCalculator) {
         super(scoreDirectorFactory, constraintMatchEnabledPreference);
+        if (constraintMatchEnabledPreference
+                && !(incrementalScoreCalculator instanceof ConstraintMatchAwareIncrementalScoreCalculator)) {
+            throw new IllegalStateException("The score calculator " + incrementalScoreCalculator + " must implement "
+                    + ConstraintMatchAwareIncrementalScoreCalculator.class.getSimpleName()
+                    + " when constraintMatchEnabledPreference is true.");
+        }
```

Comment:
There's a reason why constraintMatchEnabledPreference is called "preference" (we might want to add a comment why that is). It's like a wish, a request: We prefer to have constraints match enabled (and are ready to pay the perf impact for that - it's for the UI explaining the score for example or a demanded supply of 1 of the algo's - most algo's don't demand it), but if the score director impl doesn't support it, then we don't get it anyway.